apps/api/src/hyp3_api/validation.py

```diff
@@ -70,31 +70,48 @@ def check_granules_exist(granules, granule_metadata):
         raise GranuleValidationError(f'Some requested scenes could not be found: {", ".join(not_found_granules)}')
 
 
-def check_dem_coverage(granule_metadata):
+def check_dem_coverage(_, granule_metadata):
     bad_granules = [g['name'] for g in granule_metadata if not has_sufficient_coverage(g['polygon'])]
     if bad_granules:
         raise GranuleValidationError(f'Some requested scenes do not have DEM coverage: {", ".join(bad_granules)}')
 
 
-def check_same_burst_ids(granule_metadata):
-    ref_burst_id, sec_burst_id = [granule['name'].split('_')[1] for granule in granule_metadata]
-    if ref_burst_id != sec_burst_id:
+def check_same_burst_ids(job, _):
+    refs = job['job_parameters']['reference']
+    secs = job['job_parameters']['secondary']
+    ref_ids = [ref.split('_')[1] for ref in refs]
+    sec_ids = [sec.split('_')[1] for sec in secs]
+    if len(ref_ids) != len(sec_ids):
+        error_message = (
+            f'Number of reference and secondary scenes must match, got: '
+            f'{len(ref_ids)} references and {len(sec_ids)} secondaries'
+        )
+        raise GranuleValidationError(error_message)
+    if ref_ids != sec_ids:
+        bad_id_index = [ref_ids.index(ref) for (ref, sec) in zip(ref_ids, sec_ids) if ref != sec][0]
+        raise GranuleValidationError(
+            f'Burst IDs do not match for {refs[bad_id_index]} and {secs[bad_id_index]}.'
+        )
```

Comment:
This is the obvious way to do it, so takes less time to understand, in my opinion.
```suggestion
    for i in range(len(ref_ids)):
        if ref_ids[i] != sec_ids[i]:
            raise GranuleValidationError(
                f'Burst IDs do not match for {refs[i]} and {secs[i]}.'
            )
```

Reply:
I've been staring at so many list comprehensions that that somehow didn't occur to me lol

Reply:
My only complaint would be that it runs the loop every time, but I guess these are all tiny lists so it shouldn't matter.

Reply:
But keep in mind that `ref_ids != sec_ids` is also a looping operation, as is `ref_ids.index(ref)`, plus the list comprehension. So you've actually got three loops here :)